### ISSUE TITLE
Render Reform

### ISSUE THREAD

**Issue by excaliburHisSheath** [excaliburHisSheath/gunship-rs#65]
The core goal here is to refactor polygon to make it more self-contained. Currently it acts as a thin wrapper around OpenGL, handing off ownership of OpenGL resources like meshes and shader programs to the engine and requiring the engine to pass those objects back in order to draw. Instead we'd like polygon to hold onto backend-specific details and manage all rendering processes internally. The broad strokes are as follows:

- [x] Don't use raw OpenGL, focus on higher-level structure and functionality and leave the trouble of dealing with the fiddly OpenGL implementation details to gl-util (depends on #64).
- [x] Polygon should manage the positions of its various renderable objects internally rather than having to have the transform values passed in when drawing.
- [x] Don't expose backend-details like `GLMesh`. Instead manage those details internally and only give unique identifiers that can be used independent of backend.
- [x] Use a basic material system that allows shaders and parameter values to be specified for a mesh when drawing.

# Remaining Questions

- Does polygon keep a full scene graph or a flat list of drawables? I like the idea of polygon not having to worry about parent-child relationships and only focussing on drawing an unordered set of objects but will that cause problems with things like culling and bounding boxes that may be able to take advantage of parent/child relationships?
- For things like mesh and shader identifiers what should we use to allow client code to manage those values? Should we just pass back unique identifiers the way that OpenGL does (but make them backend agnostic)? Should we make those objects "heavier" by making them non-cloneable (or reference counting if they are clonable) and using RAII to automatically cleanup resources when they're done?

**Comment by excaliburHisSheath** [comment]
This work is currently in progress on the [`render-reform`](https://github.com/excaliburHisSheath/gunship-rs/tree/render-reform) branch.

**Comment by excaliburHisSheath** [comment]
As of 8009a375ac12068fcc5893c686048a5bd72abfc8 this work is complete, it's just waiting on some other related work before it can be merged into `master`.
